submariner/multiple-cable-drivers.md

```diff
@@ -0,0 +1,70 @@
+## Support multiple cable drivers within the same supercluster
+
+## Summary
+
+Submariner supports several cable driver implementations (libreswan, wireguard et al) however only one can be 
+configured within the same supercluster, that is all clusters that join a broker must use the same cable driver.
+
+There are use cases that involve supporting a mix of cable drivers or cable driver configurations within the same
+supercluster. For example, a supercluster could contain three clusters, two that are on-premise and the other
+in the cloud. Connections to the cloud cluster will need to use encryption while the on-premise clusters can use
+unencrypted connections, as they’re connected by a private link, to avoid the performance overhead of encryption.
+This could be achieved using different cable driver implementations or the same cable driver implementation with
+differing encryption configurations.
+
+## Proposal
+
+### User-facing API
+
+There needs to be a way for a user to specify their intent for two clusters to use a specific cable driver
+implementation and/or configuration. At a minimum, the intent must include identifying information for the two
+clusters and the cable driver implementation.
+
+The standard mechanism in Kubernetes to specify intent is by creating a Kubernetes API resource whose effect is eventually
+realized into the cluster's desired state. To extend the Kubernetes API, one could either use the generic Kubernetes
+`ConfigMap` or use a custom resource definition (CRD). There's pros and cons to each but a CRD seems more suitable for
+this enhancement.
+
+This following CRD is proposed:
+
+```
+appVersion: v1
+kind: ClusterLinkConfig
+metadata:
+  name: east-west-link
+spec:
+  leftCluster: 
+    id: east
+  rightCluster: 
+    id: west
+  cableDriver: 
+    name: libreswan
+    options:
+      encrypt: true
+```
+
+The `leftCluster` and `rightCluster` fields identify the two clusters. The simplest identifier is the cluster's unique
+ID or name within the supercluster. This could be extended in the future to allow for a label selector to identify one 
+or more Submariner `Cluster` resources.
+
+While semantically the identifiers for the two clusters uniquely identify`ClusterLinkConfig` instances, the
+`leftCluster` and `rightCluster` field values could be reversed by the user in which case two instances semantically
```

Comment:
Just thinking out loud for other ideas, what is your opinion on the following approach?

We associate each cluster with a new parameter which specifies if encryption is mandatory or optional?
This can go as part of ClusterCRD configuration itself.
An OnPrem cluster can be configured with "encryption: optional", whereas a public cluster will be configured with "encryption: required".
If an OnPrem1 cluster is trying to connect to OnPrem2 cluster, since both of them have Optional encyption, they can use non-encrypted tunnels.
However, if one of the sides have "encryption: required", then encrypted tunnels would be used.
This looks a bit simple, but I'd like to hear views on this.

CC: @nyechiel @skitt @mangelajo @mkolesnik @vthapar 

![Screenshot from 2020-07-28 14-07-17](https://user-images.githubusercontent.com/6574656/88640700-df3db580-d0db-11ea-9d2c-bb4432b7c2ba.png)

Reply:
Eventhough this is simple, we will have less flexibility in certain scenarios. For example, if we have two public clusters (say in the same Region) and we want to avoid encryption between the two clusters, it would not be possible in this approach as we would have configured "Encryption: Required" for both of them. The proposal from @tpantelis would be more flexible.

Reply:
Yes, this is really a link property, not something we can derive from cluster properties. At some point we might even need to be able to support multiple links between clouds, and some might be encrypted and others not; this is just an example, but it’s a typical HA requirement in real-world deployments.